bokehjs/src/coffee/models/plots/plot_canvas.coffee

```diff
@@ -448,14 +448,20 @@ class PlotCanvasView extends Renderer.View
         @canvas.get("canvas_height") != height)
       @canvas_view.set_dims([width, height], trigger=false)
 
-    @canvas_view.render(force_canvas)
-
     for k, v of @renderer_views
       if not @range_update_timestamp? or v.set_data_timestamp > @range_update_timestamp
         @update_dataranges()
         break
 
     ctx = @canvas_view.ctx
+    ctx.save()  # Save default state
+
+    # Set hidpi-transform
+    ratio = @canvas_view.render(force_canvas)
```

Comment:
It's unusual for `render` to return anything. But more than that, this ratio is only conditionally returned on one side of branch inside `render`, which is both hard to reason about and potentially fragile. I think this ratio should be communicated in a different manner. E.g., an internal property set on canvas init, a method that can be called, or just calling `get_scale_ratio(@ctx, @ctx.get('use_hidpi'))` here explicitly.